New optimized general convex path shadow algorithm

flar opened the pull request:
<!--
Thanks for filing a pull request!
Reviewers are typically assigned within a week of filing a request.
To learn more about code review, see our documentation on Tree Hygiene: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
-->

Convex paths will now use an optimized mesh-based algorithm to render shadows. The algorithm is based on the code in SkShadowTessellator.

Fixes https://github.com/flutter/flutter/issues/170764

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

flar committed:
New optimized general convex path shadow algorithm

flar committed:
Merge branch 'master' of github.com:flutter/flutter into general-convex-path-shadows

flar committed:
more consistent types in new shaders and functions

flar committed:
restore inadvertently commented out interop test

flar committed:
mallioc

flar committed:
fix nop-blur detection

flar committed:
fix comment

flar committed:
adjust shadow radius to match other algorithms

flar committed:
adjust comments

flar committed:
remove Skia shadow mesh hook used for debugging

flar committed:
improve value and coverage of the playground tests

flar committed:
Merge branch 'master' of github.com:flutter/flutter into general-convex-path-shadows

flar commented:
It looks like the new Blur contents overrides Render instead of SetPassInfo. I need to check if it is bypassing some of the setup in the parent class.

gaaclarke reviewed (commented):
I'll hold off reviewing until the goldens are good.  Ping me when you're ready.

flar committed:
fix Inner/Outer mask blurs with new general convex shadow rendering

flar committed:
switch to local coordinates for compatibility with various entity configurations

flar committed:
Merge branch 'master' of github.com:flutter/flutter into general-convex-path-shadows

flar commented:
Updating to head as a possible de-flake because the "smiley face android engine test" does not look related and doesn't reproduce locally.

flar commented:
The new goldens all make sense to me so this is ready for review now.

gaaclarke commented on engine/src/flutter/impeller/display_list/aiks_dl_shadow_unittests.cc:
```diff
@@ -0,0 +1,869 @@
+TEST_P(AiksTest, DrawShadowDoesNotOptimizeHourglass) {
+  DisplayListBuilder builder;
+  builder.Clear(DlColor::kWhite());
+  builder.Scale(GetContentScale().x, GetContentScale().y);
```
I thought we got rid of the need to do this, did you find you had to do it?

gaaclarke replied:
Ugh, wonder what happened there, dreamtime coding session where I fixed all my annoyances?

gaaclarke commented on engine/src/flutter/impeller/entity/geometry/shadow_path_geometry.cc:
```diff
@@ -0,0 +1,1264 @@
+    // polygon. The final list of vertices that remain in the umbra polygon
+    // are the vertices that remain on this linked list from a "head" pin.
+    UmbraPin* pNext = nullptr;
+    UmbraPin* pPrev = nullptr;
```
Can't this be inferred from the index in the vector?

gaaclarke commented on engine/src/flutter/impeller/entity/geometry/shadow_path_geometry.cc:
```diff
@@ -0,0 +1,1264 @@
+  // |p_curr_pin| starting from the absolute point |fan_start| and ending
+  // at the absolute point |fan_end|, both of which should be equi-distant
+  // from the path vertex. The index of the vertex at |fan_start| should
+  // already be in the vector of vertices at an index given by |start_index|.
```
nit: This is in Dart docstring format.  I would prefer the style we use in the engine which documents the arguments in their own section and what is returned.

gaaclarke commented on engine/src/flutter/impeller/entity/geometry/shadow_path_geometry.cc:
```diff
@@ -0,0 +1,1264 @@
+  if (!matrix.IsInvertible()) {
+    Invalidate();
+    return;
+  }
```
Instead of creating invalid objects, please use absl::StatusOr with a factory function.

gaaclarke commented on engine/src/flutter/impeller/entity/geometry/shadow_path_geometry.cc:
```diff
@@ -0,0 +1,1264 @@
+  Vector2 tip_delta = pPin1->pin_tip - pPin0->pin_tip;
+  Vector2 w = tip_delta;
+  Scalar denom = pPin0->path_delta.Cross(pPin1->path_delta);
+  bool denomPositive = (denom > 0);
```
```suggestion
  bool denom_positive = (denom > 0);
```

gaaclarke commented on engine/src/flutter/impeller/entity/geometry/shadow_path_geometry.h:
```diff
@@ -0,0 +1,108 @@
+
+  const std::vector<Point>& GetVertices() const { return vertices_; }
+  const std::vector<uint16_t>& GetIndices() const { return indices_; }
+  const std::vector<Scalar>& GetGaussians() const { return gaussians_; }
```
Or maybe that comment would be better here.

gaaclarke commented on engine/src/flutter/impeller/entity/geometry/shadow_path_geometry_unittests.cc:
```diff
@@ -0,0 +1,344 @@
+namespace impeller {
+namespace testing {
+
+using namespace flutter;
```
using namespace is technically forbidden, can you list the symbols you need?

flar replied:
Most of the unit test files use the entire flutter namespace.

We even have implementation files that do this as well. :(

gaaclarke replied:
leave the campsite cleaner that you found it =)

gaaclarke commented on engine/src/flutter/impeller/entity/geometry/shadow_path_geometry_unittests.cc:
```diff
@@ -0,0 +1,344 @@
+#include "flutter/third_party/skia/src/core/SkVerticesPriv.h"  // nogncheck
+#include "flutter/third_party/skia/src/utils/SkShadowTessellator.h"  // nogncheck
+
+#define SHOW_VERTICES false
```
```suggestion
#define IMPELLER_SHOW_VERTICES false
```

Please add a prefix to avoid any collisions.

gaaclarke commented on engine/src/flutter/impeller/geometry/path_source.h:
```diff
@@ -109,6 +109,38 @@ class EllipsePathSource : public PathSource {
+
+ private:
+  PathReceiver& receiver_;
+  const impeller::Matrix& matrix_;
```
Raw pointers.

gaaclarke reviewed (changes requested):
Woo, I'm going to need a nap now.  I wish there was a way to make these PRs smaller in the future =). Looking good for the most part but it's a big PR so there was a lot to go over.

flar committed:
Merge branch 'master' of github.com:flutter/flutter into general-convex-path-shadows

flar commented on engine/src/flutter/impeller/display_list/aiks_dl_shadow_unittests.cc:
```diff
@@ -0,0 +1,869 @@
+}
+}  // namespace
+
+TEST_P(AiksTest, DrawShadowDoesNotOptimizeHourglass) {
```
It gives a view that shows how the mesh is formed for edification purposes only. Also, at one point (if you look back in the commits) I was also creating meshes using the equivalent Skia call to see if our 2 versions of the algorithms agreed.

They can be eliminated if we don't want them.

flar replied:
One thing to note, it shows how the mesh differs between the CW and CCW versions of the path (the diagonal lines on the quads reverse).

flar replied:
I don't want to assert the values of the mesh because it can vary depending on platform (Intel vs ARM). I do assert the number of vertices in the other test file at least, and I had to adjust the numbers in one of those tests because Linux was producing one fewer vertex on the corner fans so I made the numbers not be so close to the cutoffs.

flar replied:
This is pretty optional, but it helps to verify that our meshes "look right".

flar commented on engine/src/flutter/impeller/display_list/canvas.h:
```diff
@@ -283,31 +283,17 @@ class Canvas {
   };
+  class RRectBlurShape;
+  class RSuperellipseBlurShape;
+  class PathBlurShape;
```
There is a private method `AttemptDrawBlur` that takes an instance of this class. It is internal only and should not leak outside this class.

Originally these 3 classes were defined inline in the header, but I moved them to the cc file to remove implementation from the header. They cannot implement this private interface unless they are mentioned in the class's (private) definition, so I needed the forward declaration.

So, implementation in header file?
Or publicly leaked internal interface?
Or forward declaration?

flar commented on engine/src/flutter/impeller/display_list/canvas.h:
```diff
@@ -395,22 +381,24 @@ class Canvas {
-                               const Paint& paint);
+  /// Returns the radius common to both width and height of all corners,
+  /// or -1 if the radii are not uniform.
+  Scalar GetCommonRRectLikeRadius(const RoundingRadii& radii);
```
`IsShadowBlurDrawOperation` was the only other one I could find that could be const or static and I made it static.

flar commented on engine/src/flutter/impeller/entity/contents/shadow_vertices_contents.h:
```diff
@@ -0,0 +1,51 @@
+class ShadowVerticesContents final : public SolidBlurContents {
+ public:
+  static std::shared_ptr<ShadowVerticesContents> Make(
+      const ShadowPathGeometry* geometry);
```
That seems to work...

flar commented on engine/src/flutter/impeller/entity/geometry/fill_path_geometry.h:
```diff
@@ -49,6 +49,21 @@ class FillPathSourceGeometry : public Geometry {
+ public:
+  explicit FillPathFromSourceGeometry(
+      const PathSource& source,
+      std::optional<Rect> inner_rect = std::nullopt);
```
Unused, deleted.

gaaclarke commented on engine/src/flutter/impeller/compiler/shader_lib/impeller/gaussian.glsl:
```diff
@@ -78,4 +78,17 @@ float16_t IPSigmoid(float16_t x) {
 
+/// Converts a fraction in the range [0,1] to a guassian weighted distribution
+/// over the same range ([0,1]).
+float16_t IPFractionToGaussian(float16_t fraction) {
```
Did you forget to push your PR?  I don't know if the thumbs up means "i did it" or "good idea, let's do it" haha.

flar commented on engine/src/flutter/impeller/compiler/shader_lib/impeller/gaussian.glsl:
```diff
@@ -78,4 +78,17 @@ float16_t IPSigmoid(float16_t x) {
+/// Converts a fraction in the range [0,1] to a guassian weighted distribution
+/// over the same range ([0,1]).
+float16_t IPFractionToGaussian(float16_t fraction) {
+  // IPErf produces outputs over [0, 1] from an input range of [-2, +2].
```
I also mapped it and the function appears to do the right thing with the way I call it. I adjust [0,1] to [-2,+2] and get the right answers out of it. Here was my graph of the math: https://www.desmos.com/calculator/i4kuidzal5

flar commented on engine/src/flutter/impeller/entity/geometry/shadow_path_geometry.cc:
```diff
@@ -0,0 +1,1264 @@
+}
+
+// Adjust the device point to its nearest sub-pixel grid location.
+Point PolygonInfo::ToDeviceGrid(Point point) {
```
static.

flar commented on engine/src/flutter/impeller/entity/geometry/shadow_path_geometry.cc:
```diff
@@ -0,0 +1,1264 @@
+  x_direction_detector_.AccumulateDirection(edge_vector.x);
+  y_direction_detector_.AccumulateDirection(edge_vector.y);
+  return x_direction_detector_.change_count <= 3u &&
+         y_direction_detector_.change_count <= 3u;
```
I moved the test into the struct and commented on why 3 makes sense.

flar replied:
This isn't a tunable characteristic where you might want to change the value it compares against, it is a mathematical axiom so I think a comment on why that number makes sense at the point where we implement the new `IsConvex()` function on the accumulator struct does a better job than simply pulling it out into a separate constant which would also move its definition away from where it is applied.

flar commented on engine/src/flutter/impeller/entity/geometry/shadow_path_geometry.cc:
```diff
@@ -0,0 +1,1264 @@
+// naming and differences in the methods on Point and Vertex2.
+std::optional<PolygonInfo::PinIntersection> PolygonInfo::ComputeIntersection(
+    UmbraPin* pPin0,
+    UmbraPin* pPin1) {
```
Already changed when I adjusted the naming and associations with the return value fractions.

flar commented on engine/src/flutter/impeller/entity/geometry/shadow_path_geometry.cc:
```diff
@@ -0,0 +1,1264 @@
+// This method was converted nearly verbatim from the Skia source files
+// SkShadowTessellator.cpp and SkPolyUtils.cpp, except for variable
+// naming and differences in the methods on Point and Vertex2.
+std::optional<PolygonInfo::PinIntersection> PolygonInfo::ComputeIntersection(
```
Not error handling at all. optional is exactly the right answer here as it is answering both "is there an intersection and if so what are its parameters".

flar commented:
I've fixed pretty much all of the minor stuff, and the main things left are related to lifetimes of refs and ptrs which I want to rethink. I'm going to push a new commit that gets the minor stuff out of the way while I think about that. I'll leave it to @gaaclarke to "resolve" the comments that I've addressed to his satisfaction.

flar committed:
gaaclarke feedback

flar committed:
check for overlapping triangles and simplify PathBlurShape interactions

flar committed:
track MeshStatus more deliberately

flar committed:
document MeshStatus

flar committed:
Merge branch 'master' of github.com:flutter/flutter into general-convex-path-shadows

flar committed:
swap lots of fields for local variable and greatly reduce scope of remaining status fields

flar committed:
lots of comments/tests and separate path accumulation/deduplication from convexity and centroid analysis

flar committed:
adjust test data to not trigger bit-error edge cases on different platforms

flar committed:
Merge branch 'master' of github.com:flutter/flutter into general-convex-path-shadows

gaaclarke commented on engine/src/flutter/impeller/display_list/aiks_dl_shadow_unittests.cc:
```diff
@@ -17,46 +17,57 @@ namespace testing {
 
   // |PathReceiver|
-  virtual void MoveTo(const Point& p2, bool will_be_closed) {
+  virtual void MoveTo(const Point& p2, bool will_be_closed) override {
```
nit: "virtual" is redundant once you have "override"

gaaclarke commented on engine/src/flutter/impeller/entity/geometry/shadow_path_geometry.cc:
```diff
@@ -21,50 +21,211 @@ using impeller::Tessellator;
 using impeller::Vector2;
 
+// Each point in the polygon form of the path is turned into a structure
+// that tracks the gradient of the shadow at that point in the path. The
```
These should be doxygen comments (3 forward slashes)

gaaclarke commented on engine/src/flutter/impeller/display_list/canvas.cc:
```diff
@@ -689,8 +813,20 @@ void Canvas::DrawOval(const Rect& rect, const Paint& paint) {
-    return;
+  if (IsShadowBlurDrawOperation(paint)) {
+    if (rect.IsSquare()) {
+      RRectBlurShape shape(rect, rect.GetWidth() * 0.5f);
```
Yea, or a named helper function.  There is no context here that communicates you are converting from a diameter to a radius.

gaaclarke commented on engine/src/flutter/impeller/entity/geometry/shadow_path_geometry.cc:
```diff
@@ -0,0 +1,1264 @@
+    }
+
+    // Check for zero-length segments
+    Scalar v0dotv0 = FiniteVectorLengthSquared(v0);
```
I mentioned it because there is indexing in the name.  The feedback would be more applicable to the v0 variable.  It wasn't apparent to me that the name of the variable is just the operation that was used to derive it.  Usually we name things what they represent, not how they were calculated.  I'd prefer `v0_length_squared`.

gaaclarke reviewed (commented):
This is looking really close, thanks @flar.  There are a few straggling comment issues and naming issues but otherwise is looking good.  Can you do me a favor and respond to my issues that I haven't marked "resolved".  When there is a big review it's helpful to get some feedback that a comment has been addressed, even if it just says "done".  Thanks.

flar commented on engine/src/flutter/impeller/entity/geometry/shadow_path_geometry_unittests.cc:
```diff
@@ -0,0 +1,344 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
```
I found that one of the biggest blocks in the logic cannot be executed because it is implemented in a "general polygon" module that doesn't realize that the polygon was normalized by the calling code. I haven't proven that it can no longer be executed, so I added a big comment about the issue to the block. I plan to come back and delete the block later when I explore the 2 methods that I left "as is" from the Skia code.

flar committed:
review feedback

gaaclarke reviewed (approved):
lgtm, thanks jim

flar committed:
fix stitching of shadows with degenerate umbra polygons

flar commented:
I reran my blur size comparison tests and I'm happy with the consistency of the output between the optimized special-case maskblur implementations (the old AttemptRRectBlurShape which is still used for basic rects and rrects), the general case maskblur (i.e. uses a GaussianBlur ImageFilter for any kind of shape) and the new general convex path shader (this PR).

flar committed:
Merge branch 'master' of github.com:flutter/flutter into general-convex-path-shadows

flar commented on engine/src/flutter/impeller/display_list/aiks_dl_shadow_unittests.cc:
```diff
@@ -99,18 +99,28 @@ void DrawShadowMesh(DisplayListBuilder& builder,
+
+  // From dl_dispatcher, making a MaskFilter.
+  Scalar light_radius = 800 / 600;
+  EXPECT_EQ(light_radius, 1.0f);  // Value in dl_dispatcher is bad.
```
Yep. I should file a separate issue. It's been there since that code was first written.
